docs/source/more_types.rst

```diff
@@ -178,6 +183,357 @@ create subclasses of these objects either.
+
+    from typing import overload, Union
+
+    class A: pass
+    class B: pass
+
+    @overload
+    def func(x: A) -> int: ...
+
+    @overload
+    def func(x: B) -> str: ...
+
+    def func(x: Union[A, B]) -> Union[int, str]:
+        if isinstance(x, A):
+            return 42
+        else:
+            return "foobar"
+
+    def should_return_str(x: B) -> str:
+        return func(x)
+
+This program is fine under most normal conditions. However, if we
+create a new class that inherits from *both* A and B, we get the
+same kind of discrepency we had earlier:
+
+.. code-block:: python
+
+    class Combined(A, B): pass
+
+    # bad_var is of type 'str' according to mypy but is in reality of type 'int'.
+    bad_var = should_return_str(Combined())
+
+Mypy could prohibit this by flagging the definition of ``func``
+as being unsafe since ``A`` and ``B`` could theoretically overlap
+due to multiple inheritance. However, multiple inheritance is
+uncommon and making such a change would make it difficult
+to use overloads in a useful way. Consequently, mypy is designed to
+ignore the possibility of multiple inheritance altogether when
+checking overloads: it will *not* report any errors with the above program.
+
+Thankfully, these types of situations are relatively rare. What this
+does mean, however, is that you should exercise caution when using an
+overloaded function in scenarios where it can potentially receive values
+that are an instance of two seemingly unrelated types. For example,
+``Combined()`` is an instance of both ``A`` and ``B``; the empty list 
+``[]`` is an instance of both ``List[int]`` and ``List[str]``.
+
+
+Type checking the implementation
+--------------------------------
+
+The body of an implementation is type-checked against the
+type hints provided on the implementation. For example, in the
+``MyList`` example up above, the code in the body is checked with
+argument list ``index: Union[int, slice]`` and a return type of 
+``Union[T, Sequence[T]]``. If there are no annotations on the
+implementation, then the body is not type checked. If you want to
+force mypy to check the body anyways, use the ``--check-untyped-defs``
+flag (:ref:`more details here <additional-command-line-flags>`).
+
+The variants must also also be compatible with the implementation
+type hints. In the ``MyList`` example, mypy will check that the
+parameter type ``int`` and the return type ``T`` are compatible with
+``Union[int, slice]`` and ``Union[T, Sequence]`` for the
+first variant. For the second variant it verifies the parameter
+type ``slice`` and the return type ``Sequence[T]`` are compatible
+with ``Union[int, slice]`` and ``Union[T, Sequence]``.
+
+.. note::
+
+   The overload semantics documented above are new as of mypy 0.620.
+
+   Older versions of mypy used different semantics. In particular, overload
+   variants had to different after erasing types and calls to overloaded
+   functions did not use the "pick the first match" rule.
```

Comment:
You can mention that mypy previously often silently returned `Any` for many overload calls. So that people will be not surprised by new errors when switching to 0.620.

Reply:
Will do